### ISSUE TITLE
load(url).asGif().toBytes() vs load(url).asGif()

### ISSUE THREAD

**Issue by jaredsburrows** [bumptech/glide#1526]
## Glide.with(context).load(url).asGif().toBytes()

This code will auto-play the gif when it is done downloading but the performance is terrible.

```java
   Glide.with(context)
      .load(url)
      .asGif()
      .toBytes()
      .thumbnail(THUMBNAIL_MULTIPLIER)
      .override(GIF_IMAGE_WIDTH_PIXELS, GIF_IMAGE_HEIGHT_PIXELS)
      .error(R.mipmap.ic_launcher)
      .into(new SimpleTarget<byte[]>() {
        @Override public void onResourceReady(byte[] resource, GlideAnimation<? super byte[]> glideAnimation) {
          // Load gif
          try {
            imageView.setImageDrawable(new GifDrawable(resource));
          } catch (IOException e) {
            imageView.setImageResource(R.mipmap.ic_launcher);
          }
          imageView.setVisibility(View.VISIBLE);

          // Turn off progressbar
          progressBar.setVisibility(View.INVISIBLE);
          if (Log.isLoggable(TAG, Log.INFO)) {
            Log.i(TAG, "finished loading\t" + Arrays.toString(resource));
          }
        }
      });
```
## Glide.with(context).load(url).asGif()

This code will auto-play the gif when all the gifs are done downloading. So in a `RecyclerView`, if you scroll and download more images, they seem to "pause" but the performance is great!

```java
    Glide.with(context)
      .load(url)
      .asGif()
      .diskCacheStrategy(DiskCacheStrategy.SOURCE)
      .listener(new RequestListener<String, GifDrawable>() {
        @Override public boolean onException(Exception e, String model, Target<GifDrawable> target, boolean isFirstResource) {
          // Show gif
          imageView.setImageResource(R.mipmap.ic_launcher);
          imageView.setVisibility(View.VISIBLE);

          // Hide progressbar
          progressBar.setVisibility(View.GONE);
          if (Log.isLoggable(TAG, Log.INFO)) {
            Log.i(TAG, "finished loading\t" + model);
          }
          return false;
        }

        @Override public boolean onResourceReady(GifDrawable resource, String model, Target<GifDrawable> target, boolean isFromMemoryCache, boolean isFirstResource) {
          // Show gif
          imageView.setVisibility(View.VISIBLE);

          // Hide progressbar
          progressBar.setVisibility(View.GONE);
          if (Log.isLoggable(TAG, Log.INFO)) {
            Log.i(TAG, "finished loading\t" + model);
          }
          return false;
        }
      })
      .into(imageView);
```

How can I combine both of these strategies?
- I want it to auto play right when it finishes
- When I scroll, it should not affect the performance or stop playing

**Comment by TWiStErRob** [comment]
Sorry for the delay, your issue stuck in my queue.

The performance hit when loading an image comes from the fact that the full GIF loaded into memory, then individual frames decoded and transformed, and then all of this is encoded back into GIF. All this is implemented in Java, not native, and is required to satisfy `toBytes`. You can try `.diskCacheStrategy(SOURCE)` there too, but I think it may not have any effect, because `toBytes` forces an encoding pass anyway.

The performance hit (pause) comes from the fact that there are limited number of threads decoding the loads. When you scroll down the items that come into views all start new loads, and any still visible GIFs also keep playing their animation. With Glide's `GifDrawable` this playing happens via normal Glide loads which compete for those few threads. Each frame is decoded separately whenever needed. I assume the `GifDrawable` in the first example is an external library, and hence it does the playback independently of Glide and possibly natively.

I think you can make the second one a better experience, if you decrease the priority of the loads in your recycler view bind: `.priority(LOW)`, so the currently playing GIFs have to complete their frames first, before new images are loaded. There may be a drawback here: if the frame delay is too low (and/or there are a lot of GIFs playing), the playing GIFs may prevent the new items to bind and show their images.

**Comment by jaredsburrows** [comment]
To avoid the "pause", I implemented endless scrolling in my RecyclerView Gridlayout example. 

Here is how I do it now(https://github.com/jaredsburrows/android-gif-example/blob/master/src/main/java/burrows/apps/example/gif/data/rest/repository/ImageRepository.java#L24):

```
  public GifRequestBuilder<?> load(Object url) {
    return Glide.with(context)
      .load(url)
      .asGif()
      .thumbnail(THUMBNAIL_MULTIPLIER)
      .error(R.mipmap.ic_launcher)
      .fallback(R.mipmap.ic_launcher)
      .override(imageWidth, imageHeight)
      // https://github.com/bumptech/glide/issues/600#issuecomment-135541121
      .diskCacheStrategy(DiskCacheStrategy.SOURCE);
  }
```

Load them in the adapter like this(https://github.com/jaredsburrows/android-gif-example/blob/master/src/main/java/burrows/apps/example/gif/presentation/adapter/GifAdapter.java#L47):

```
    repository.load(imageInfo.url())
      .thumbnail(repository.load(imageInfo.previewUrl()))
      .listener(new RequestListener<Object, GifDrawable>() {
        @Override public boolean onException(Exception e, Object model, Target<GifDrawable> target, boolean isFirstResource) {
          // Show gif
          holder.imageView.setImageResource(R.mipmap.ic_launcher);
          holder.imageView.setVisibility(View.VISIBLE);

          // Hide progressbar
          holder.progressBar.setVisibility(View.GONE);
          if (Log.isLoggable(TAG, Log.INFO)) {
            Log.i(TAG, "finished loading\t" + model);
          }
          return false;
        }

        @Override public boolean onResourceReady(GifDrawable resource, Object model, Target<GifDrawable> target, boolean isFromMemoryCache, boolean isFirstResource) {
          // Show gif
          holder.imageView.setVisibility(View.VISIBLE);

          // Hide progressbar
          holder.progressBar.setVisibility(View.GONE);
          if (Log.isLoggable(TAG, Log.INFO)) {
            Log.i(TAG, "finished loading\t" + model);
          }
          return false;
        }
      })
      .into(holder.imageView);
``` 

I have found a post where it was suggested to load a "thumbnail". I made sure I chose the lowest quality gif for the thumbnail and a normal size gif to load for the actual image. This, combined with endless scrolling has seemed to dramatically help the performance.

**Comment by TWiStErRob** [comment]
Few thoughs:
 * glad you figured things out, when you say endless scrolling do you mean a Glide preloader?
 * `holder.imageView.setImageResource` is unnecessary,  you already have an error drawable.
 * for generics have you tried?  
`public <T> GifRequestBuilder<T> load(T model)`
 * do you actually use thumbnail multiplier? You may have 3 loads going on for each item.  
is preview a still GIF image?
 * `holder.imageView` should be visible the whole time, see #1571 for example.

**Comment by jaredsburrows** [comment]
@TWiStErRob 
- I could not figure out how to get that to work. I just added a `addOnScrollListener ` here: https://github.com/jaredsburrows/android-gif-example/blob/master/src/main/java/burrows/apps/example/gif/presentation/main/MainFragment.java#L153

Let me try this advice, thanks!

**Comment by TWiStErRob** [comment]
Take a look at the [v4 RV impl](https://github.com/bumptech/glide/tree/master/integration/recyclerview/src/main/java/com/bumptech/glide/integration/recyclerview) or the [backported version](https://github.com/TWiStErRob/glide-support/blob/master/src/glide3/thirdparty/com/bumptech/glide/integration/recyclerview/RecyclerViewPreloader.java) and [usage](https://github.com/TWiStErRob/glide-support/blob/20483729ba6c17b3b2194197014714c126904890/src/glide3/java/com/bumptech/glide/supportapp/github/_1189_recycler_preload/TestFragment.java) from #1189

**Comment by jaredsburrows** [comment]
@TWiStErRob 
Based on your thoughts:
 1. I will try this.
 2. I removed - `holder.imageView.setImageResource`.
 3. Thanks for the generics - `public <T> GifRequestBuilder<T> load(T model)`.
 4. I decided to remove `.thumbnail(THUMBNAIL_MULTIPLIER)`.
 5. After doing step 2, I also made `holder.imageView` visible.

Here is the change log here: https://github.com/jaredsburrows/android-gif-example/commit/246897dfa40f594dfb2b0a697b1a24974c119dcf!
